### ISSUE TITLE
Requesting a clearly-defined NO LLM Policy

### ISSUE THREAD

**Issue by lanerussell** [element-hq/element-meta#3225]
Hello,

I'd like to talk about your AI policy. Below are some reasons why using AI can result in poorer code quality:

## Poor Code Quality

Vibe coding / agentic workflows result in poorer code quality, and relaxed oversight practices. These effects may be compounded by the common practice of using additional LLM-based tooling to provide code-reviews.

* [How I Dropped Our Production Database and Now Pay 10% More for AWS](http://archive.today/2026.03.06-144058/https://alexeyondata.substack.com/p/how-i-dropped-our-production-database)
* [Claude Tested Everything Except the One Thing That Mattered (AI agent refuses to follow explicit instructions to test `createPost()` in increasingly erratic ways)](http://archive.today/2026.03.09-201135/https://christophermeiklejohn.com/ai/claude/2026/03/08/claude-tested-everything-except-the-one-thing-that-mattered.html)
* [Amazon calls engineers for a “deep dive” internal meeting to discuss “GenAI”-related outages](https://ghostarchive.org/archive/3TfgF)
* GitClear has released reports in [2024](https://www.gitclear.com/coding_on_copilot_data_shows_ais_downward_pressure_on_code_quality) and [2025](https://www.gitclear.com/ai_assistant_code_quality_2025_research) indicating a worsening of key code quality metrics correlating with increased LLM adoption.

### Deskilling

There is increasing evidence to show that LLMs negatively impact developers' coding abilities:

* [Brains show less activity when completing tasks with LLMs](https://arxiv.org/abs/2506.08872) compared to completing tasks with search or completing tasks without digital help.
* [Developers who use early-2025 LLMs](https://metr.org/blog/2025-07-10-early-2025-ai-experienced-os-dev-study/) reported higher subjective performance, but were measured to have lower objective performance. This gap between subjective and objective performance was considered notable.
* In an [Anthropic study](https://www.anthropic.com/research/AI-assistance-coding-skills), learners using LLMs demonstrated lower learning rates on average compared to learners not using LLMs.
* [A recent study uses the term "cognitive surrender"](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=6097646) to describe the way humans tend to offload key critical thinking skills onto LLMs, even when the output is wrong.
* There are [anecdotal reports of LLM users forgetting how to code](https://www.theverge.com/ai-artificial-intelligence/767973/vibe-coding-ai-future-end-evolution).
* A paper entitled ["AI Assistance Reduces Persistence and Hurts Independent Performance"](https://arxiv.org/abs/2604.04721) from April 2026 by academics from MIT, Oxford, UCLA, and Carnegie Mellon showed alarming evidence that performing a variety of tasks with the help of AI for only 10 minutes causes "inpaired unassisted performance and reduced persistence". The researchers noted that "although AI assistance improves performance in the short-term, people perform significantly worse without AI and are more likely to give up"; they also pointed out that "these findings are particularly concerning because persistence is foundational to skill acquisition and is one of the strongest predictors of long-term learning".

### Infosec risks

LLM usage results in massive security holes.

* [Meta Security Researcher's AI Agent Accidentally Deleted Her Emails](http://archive.today/2026.02.26-153034/https://www.pcmag.com/news/meta-security-researchers-openclaw-ai-agent-accidentally-deleted-her-emails)
* [Moltbook's "vibe-coded" breach is the future of security failures](https://thehill.com/opinion/cybersecurity/5744310-ai-powered-security-risks/)
* [In a study evaluating over 500k code samples, LLM-generated code was found to contain more high-risk security vulnerabilities than human-generated code](https://arxiv.org/abs/2508.21634)
* [LLMs make up package names, making them vulnerable to incorporating malicious code in "slopsquatting" attacks](https://www.theregister.com/2025/04/12/ai_code_suggestions_sabotage_supply_chain/) ([Arxiv study](https://arxiv.org/abs/2406.10279))

## What You Can Do

I'd like to kindly suggest implementing a strict "No AI" policy. This can also help with what seems to be larger and larger influxes of low quality drive-by PRs by AI agents.

You can find other projects that have taken steps to say no to AI here:
https://noai.starlightnet.work/list.html

For a specific policy, I can recommend the following policies for inspiration:

- [wafrn's Contributring Doc](https://codeberg.org/wafrn/wafrn/src/branch/main/CONTRIBUTING.md)

- [Loupe's Use of Generative AI Policy](https://discourse.gnome.org/t/loupe-no-longer-allows-generative-ai-contributions/27327)

Like many others, I have relied on Matrix/Element services for private communication with friends and family. I hope to see continued dedication to robust encryption and privacy supported by reliable tools.

Thank you for your contributions to open and safe messaging.

**Comment by bxdxnn** [comment]
As I've said in the previous issue:

OK, if Element forbids LLMs, will you please contribute all of the future code/bug fixes/features that would be done by the Element developers and the community with the help of LLM?

Right now, we see a lot of contributions coming from the community with the help of LLMs, no matter if it was vibecoding or simple questions/reviews using agents. Would you consider Element of "better quality" if we revert everything that was done with the help of LLM?

Element has a very strict review process regarding the code quality and the UI/UX. It doesn't matter if you've used AI or not, the work you're submitting must be of the expected quality.

The articles that you've linked are examples of a very blind vibecoding, and such outages or catastrophic regressions because of vibecoding are very unlikely in FOSS where the code can be seen by everyone in the world. The "Meta Security Researcher" is also a very ridiculuos situation where a person that was appointed to do the "AI safety" at Meta can't even do basic things like having an execution permission control and queues a "Stop don't do anything" prompt instead of just typing the `/stop` command.

This is also very biased towards a single "no LLMs" view: why you haven't considered and linked numerous articles about the benefits of agentic LLMs, or at least relying to some common sense? For example, there's a lot of routine/boring processes in software development that would be a perfect job for the LLM agents.

**Comment by lanerussell** [comment]
My main intent is to start a discussion here that will lead to a clear policy being published on the use of LLMs in Element/Matrix, whatever that policy may look like. I've presented some of the concerns surrounding code quality when using LLMs. Perhaps others could discuss the questionable legality of the training data scraped by LLMs, or whether code generated by LLMs can be legally copyleft- (or otherwise) licensed. Concerns around environmental impacts, ethical implications, etc.

I'm sure others will have constructive feedback about the possible benefits of LLM use. As I said, the main thing that would be great to see come out of this discussion is a clear policy.

**Comment by Susurrus** [comment]
The current policy allows them (as it doesn't explicitly state they're banned), that should be clear. Generally policies things are a restriction, they don't explicitly allow, as it works like this for many projects. It sounds like you want them to be banned. I don't personally do, but I don't know what productive discussion is possible here: the devs already said they're fine with it as-is.

**Comment by bxdxnn** [comment]
https://github.com/element-hq/element-x-ios/pull/5625 and other PRs fixes this issue.

**Comment by lumi-me-not** [comment]
For what it's worth, I believe we should be making a stance against these dehumanization machines.

So I believe we all should strive to have full bans on it.

**Comment by bxdxnn** [comment]
@lumi-me-not how forbidding LLMs makes it "more human" that the app will have less features and more bugs (and therefore less people) and the devs will have to spend much more time on all of that? With this logic, we can go back to the woods, as all technological innovations were dehumanizing because of automation, basuc example would be the Industrial Revolution.

And again: if Element forbids LLMs, will you please contribute manually all of the future code/bug fixes/features that would be done by the Element developers and the community with the help of LLM?

**Comment by cyrneko** [comment]
> how forbidding LLMs makes it "more human" that the app will have less features and more bugs

two questions here, let me answer them for you.

1. How will forbidding LLMs make things more human?
   Every contribution is a new person. Well, often at least it is. A new person interested in making the software better, in improving some part of it, that requires trust and honest exchange of ideas to arrive at something that everyone benefits from. It is an act of human connection, is what I'm trying to say, not just "here, my code, take it or leave it". With AI-generated PRs (or "assisted" PRs), the people that submitted the PRs (if they're even people at all) often don't get back to follow-up questions to the PR, and even if they do, responding with an LLM cuts out the connection between contributor and maintainers/developers.

2. "The app will have less features and more bugs"
   We used to write features and bugfixes without AI, too. It's not like this has only become possible since ChatGPT launched. You offer no evidence for this being the case.

---

> And again: if Element forbids LLMs, will you please contribute manually all of the future code/bug fixes/features that would be done by the Element developers and the community with the help of LLM?

Element will continue working on Element regardless. Bugfixes and community contributions will also continue to exist, as they did before LLMs came around. Matrix is a community and ecosystem after all, made up of *people*.

**Comment by bxdxnn** [comment]
> the people that submitted the PRs (if they're even people at all) often don't get back to follow-up questions to the PR

This case is already forbidden in all repos, at least de-facto. PR will be closed and/or ignored if it's just unmaintainable vibecoded slop without much elaboration.

> responding with an LLM cuts out the connection between contributor and maintainers/developers.

AFAICT it's also already forbidden in Element/Matrix repos to use LLM for comms.

> 2\. We used to write features and bugfixes without AI, too. It's not like this has only become possible since ChatGPT launched. You offer no evidence for this being the case.

We also used to do plenty of stuff without technologies, like walking, talking, crafting... so what?
Yes, of course, it was possible to do it without AI, if you had a huge budget to hire a large busy team. Now we can do it much cheaper/faster yet with the same quality (depends on your knowledge/skill obviously, like it was pre-AI boom).
Why have nobody written the [most](https://github.com/element-hq/element-meta/issues/2694) [requested](https://github.com/element-hq/element-meta/issues/3201) [features](https://github.com/element-hq/element-x-android/issues/2424) and bugfixes "before ChatGPT" then if we used to do it before? Why I (and other people on my team) were the only people that did it, with the help of LLMs? Yes, it's possible to do it without LLMs, just like it's possible to do it without an IDE, LSP, `git`, `grep`, Linux, Internet, etc, but would you bother yourself so much to do it like that? And for what?

BTW, I also don't understand the "unethical" logic: was it an "unethical" act that I've submitted plenty of features/fixes using LLMs?

> Element will continue working on Element regardless. Bugfixes and community contributions will also continue to exist, as they did before LLMs came around. Matrix is a community and ecosystem after all, made up of _people_.

It's not guaranteed that the community wouldn't simply burn out or the project runs out of money due to: point one, due to: slower development, due to: banning the new faster and convenient way of writing code. Please see XMPP as the example of both 2 points (not LLM ones), you can guess that there's a reason why we're building Matrix instead of just using XMPP.

**Comment by lumi-me-not** [comment]
Walking, talking and crafting do not have these huge ethical issues associated with them. The same can be said about IDEs, LSPs and other tooling. These just aren't the same thing.

None of these things:
 - Have huge resource and power usage
 - Are created with the goal of replicating human creative outputs; in doing so they devalue legitimate human creations
 - Are associated with fascist and eugenicist movements
 - Have been trained on massive amounts of data without consent
 - Require huge datacenters, which are largely built in unconsenting communities
 - Foster dependence on big, external corporations
 - Cause skill atrophy and cognitive debt
 - Are associated with the concept of "AI" psychosis
 - Have caused suicides, stalking, murders, ..
 - Are used to create deepfakes
 - Are used to replicate other people's voices, for the purpose of extortion
 - Disproportionally affect women and minorities
 - Are built on top of exploitation and neocolonialist practices
 - Create subtle bugs

Some of these aren't directly associated with using GenAI for programming, but I object to generative models more broadly. We should not be accepting them in the first place, for any use.

Some of these are potentially fixable, yes, but the (in my opinion) most severe one of it being designed to soullessly regurgitate human creativity and ingenuity is something that can never be fixed. A lot of the others are impossible to fix in the current socioeconomic and political climate.

So, the only ethical option is to ban it all, for all professions and all purposes, and join movements against it, so it can be eliminated as much as practically possible.

Potential short-term speed gains do not justify all this harm, exploitation and abuse. There is no way to responsibly use it. This shiny new toy is not worth the massive harms.

**Comment by bxdxnn** [comment]
@lumi-me-not again, what you've listed can be applied to any technology in any extent, for example the Internet/Linux can fit most of these categories. Mind explaining how the Internet doesn't causes every point you've listed?

Do you really think LLMs were designed/created with the "soullessly regurgitate human creativity and ingenuity" objective in mind?

>Potential short-term speed gains do not justify all this harm, exploitation and abuse. There is no way to responsibly use it. This shiny new toy is not worth the massive harms.

Also could have and always can be applied to any technology when it's released, yet you would still use it (or you would also protest Internet, cars, Linux, urbanization, etc?) after the hype goes away and it becomes a normal thing.

**Comment by cyrneko** [comment]
Anyways, to bring this discussion back to something productive, there has not been official communication from someone with the authority to speak on this issue from Element. The author, and everyone else in that regard, did not come here to hold a discussion with you and instead requested clarification from project maintainers.

Our back and forth here will lead to nothing, until someone with authority to speak on this discussion actually arrives to do so.

I will be retreating from the thread until that happens, and to soothe your heart full of rage, I recommend you do so as well.

**Comment by bxdxnn** [comment]
> until someone with authority to speak on this discussion actually arrives to do so.

There've been a few PRs already that fully clarifies the LLM policy, but this issue wasn't linked and the maintainers doesn't close this.

**Comment by MixusMinimax** [comment]
I would also love a clear No-LLM policy.

The advantages of LLMs are all in quantity, not quality. You are building technical debt.

There are no more "advantages" than that. The reasons no articles in favor of LLMs were linked is because, well, frankly there are no more valid points. I urge anyone to list another advantage apart from the "amount" of contributions, which I personally don't even see as an advantage.

And the burden of proof is high here. The problems with LLMs are so huge, and provably true, that someone would have to find quite a sizeable NEED for AI. What good technology bleeds money in a speculative bubble at an unprecedented rate? How do you think they plan to make that money back? By making everyone dependent, by mandating the use of their tech through government lobbying (=bribing), and by spreading propaganda about it being "the future" and "inevitable" and "everybody likes it". The entire AI industry is a scam, and that's ignoring all the war crimes and theft. You really need to make some STRONG arguments.

In lack of a clear NEED for LLMs, you would need to prove _every_ point against LLMs wrong. Simply saying "it speeds up development" is not enough, if it were even true.

Since FOSS development worked fine before LLMs even existed, proving this will be an impossible task.

> will you please contribute all of the future code/bug fixes/features [...]

The only contributions this project would lose are contributions from new, inexperienced people, by banning AI.
By _not_ banning AI, the project would lose contributions from experienced people with human-centric values.

There is no middle ground, you can't have both, you either lose one group or the other. I think the choice is 
pretty trivial.

The main (and only) goal of matrix is to offer a human-centric service, free from government corruption, free from corporate greed. Look at what the industry does, and do the opposite, and you'll be on the right side.

I eagerly await official feedback. Reject convenience.

**Comment by bxdxnn** [comment]
>Element is mostly using Claude right now.

https://github.com/element-hq/synapse/pull/20008#discussion_r3658423061

**Comment by bxdxnn** [comment]
@MixusMinimax 

> By _not_ banning AI, the project would lose contributions from experienced people with human-centric values.

I don't see any contributions (PRs or even some basic bug/feature reports) by you (or anyone who was just posting "ban LLM" agenda here) in any Element repo.

It looks like all current Element external contributors are fine with LLM usage, and employees are actively using it. Why people that advocate anti-LLM policy doesn't have to do much with the projects where they're advocating it?

**Comment by cyrneko** [comment]
Probably because those people don't just care about the individual software that they use, but the wider open-source ecosystem and its health, which I find a noble goal that's worthy ot being supported.
